File: src/main/java/com/example/order_management/application/services/OrderService.java
package com.example.order_management.application.services;

import com.example.order_management.application.ports.in.CreateOrderUseCase;
import com.example.order_management.application.ports.in.GetOrderUseCase;
import com.example.order_management.application.ports.in.PayOrderUseCase;
import com.example.order_management.application.ports.out.OrderRepository;
import com.example.order_management.domain.exception.DomainException;
import com.example.order_management.domain.model.Money;
import com.example.order_management.domain.model.Order;
import com.example.order_management.domain.model.OrderItem;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Application service implementing order use cases.
 */
@Service
public class OrderService implements CreateOrderUseCase, GetOrderUseCase, PayOrderUseCase {
    
    private final OrderRepository orderRepository;
    
    public OrderService(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }
    
    @Override
    @Transactional
    public Order createOrder(UUID customerId, List<OrderItemRequest> items) {
        List<OrderItem> domainItems = items.stream()
            .map(item -> new OrderItem(
                item.productId(),
                item.quantity(),
                new Money(item.unitPrice())
            ))
            .collect(Collectors.toList());
        
        Order order = new Order(UUID.randomUUID(), customerId, domainItems);
        return orderRepository.save(order);
    }
    
    @Override
    @Transactional(readOnly = true)
    public Order getOrderById(UUID orderId) {
        return orderRepository.findById(orderId)
            .orElseThrow(() -> new DomainException("Order not found with ID: " + orderId));
    }
    
    @Override
    @Transactional
    public Order payOrder(UUID orderId) {
        Order order = getOrderById(orderId);
        order.markAsPaid();
        return orderRepository.save(order);
    }
}

File: src/OrderManagement.Application/Services/OrderService.cs
using OrderManagement.Application.DTOs;
using OrderManagement.Application.Exceptions;
using OrderManagement.Application.Ports;
using OrderManagement.Domain.Entities;
using OrderManagement.Domain.Enums;
using OrderManagement.Domain.Exceptions;
using OrderManagement.Domain.ValueObjects;

namespace OrderManagement.Application.Services;

/// <summary>
/// Application service that orchestrates order use cases.
/// </summary>
public sealed class OrderService : IOrderService
{
    private readonly IOrderRepository _orderRepository;
    private readonly IUnitOfWork _unitOfWork;

    public OrderService(IOrderRepository orderRepository, IUnitOfWork unitOfWork)
    {
        _orderRepository = orderRepository;
        _unitOfWork = unitOfWork;
    }

    public async Task<CreateOrderResponse> CreateOrderAsync(CreateOrderRequest request, CancellationToken cancellationToken = default)
    {
        try
        {
            var items = request.Items
                .Select(i => new OrderItem(i.ProductId, i.Quantity, new Money(i.UnitPrice)))
                .ToList();

            var order = new Order(Guid.NewGuid(), request.CustomerId, items);
            await _orderRepository.AddAsync(order, cancellationToken);
            await _unitOfWork.SaveChangesAsync(cancellationToken);

            var total = order.TotalAmount;
            return new CreateOrderResponse(
                order.Id,
                order.Status.ToString(),
                total.Amount,
                order.CreatedAt);
        }
        catch (InvalidItemException ex)
        {
            throw new ApplicationValidationException(ex.Message);
        }
        catch (CurrencyMismatchException ex)
        {
            throw new ApplicationValidationException(ex.Message);
        }
    }

    public async Task<OrderDetailResponse?> GetOrderByIdAsync(Guid orderId, CancellationToken cancellationToken = default)
    {
        var order = await _orderRepository.GetByIdAsync(orderId, cancellationToken);
        if (order is null) return null;

        var items = order.Items
            .Select(i => new OrderItemDetailResponse(
                i.ProductId,
                i.Quantity,
                i.UnitPrice.Amount,
                i.LineTotal.Amount))
            .ToList();

        var total = order.TotalAmount;
        return new OrderDetailResponse(
            order.Id,
            order.CustomerId,
            order.Status.ToString(),
            items,
            total.Amount,
            total.Currency);
    }

    public async Task<PayOrderResponse?> PayOrderAsync(Guid orderId, CancellationToken cancellationToken = default)
    {
        var order = await _orderRepository.GetByIdAsync(orderId, cancellationToken);
        if (order is null) return null;

        try
        {
            order.MarkAsPaid();
            await _orderRepository.UpdateAsync(order, cancellationToken);
            await _unitOfWork.SaveChangesAsync(cancellationToken);

            return new PayOrderResponse(order.Id, order.Status.ToString());
        }
        catch (InvalidOrderStateException ex)
        {
            throw new ApplicationConflictException(ex.Message);
        }
    }
}
